src/net/server/bytearrstream.rs

```diff
@@ -0,0 +1,452 @@
+        }
+
+        let Query(FieldStr { field_type }) = req.extract().await?;
+        let field_size = ff::size_in_bytes_from_type_str(&field_type)?;
+
+        Ok(Self { field_size })
+    }
+}
+
+/// Wraps a [`BodyStream`] and produce a new stream that has chunks of exactly size `size_in_bytes`.
+/// # Errors
+/// If the downstream body is not a multiple of `size_in_bytes`.
+#[derive(Debug)]
+#[pin_project]
+pub struct ByteArrStream {
+    #[pin]
+    body: BodyStream,
+    size_in_bytes: u32,
+    buffered_size: u32,
+    buffered: VecDeque<Bytes>,
+}
+
+impl ByteArrStream {
+    /// # Panics
+    /// if `size_in_bytes` is 0
+    pub fn new(body: BodyStream, size_in_bytes: u32) -> Self {
+        assert_ne!(size_in_bytes, 0);
+        Self {
+            body,
+            size_in_bytes,
+            buffered_size: 0,
+            buffered: VecDeque::new(),
+        }
+    }
+
+    /// returns [`Bytes`] of `size_in_bytes` length from the buffered chunks. Returns [`None`] if
+    /// there are less than `size_in_bytes` bytes in the buffer
+    fn pop_front(&mut self) -> Option<Bytes> {
+        // not enough bytes buffered
+        if self.buffered_size < self.size_in_bytes {
+            None
+        } else if self.buffered[0].len() > self.size_in_bytes as usize {
+            // if the first buffer is large enough, simply split it in-place.
+            // This is O(1) operation, and should be true majority of the time, assuming large size
+            // buffers
+            let out_bytes = self.buffered[0].split_to(self.size_in_bytes as usize);
+            self.buffered_size -= self.size_in_bytes;
+            Some(out_bytes)
+        } else {
+            // First buffer is too small, so we need to take across buffers.
+            // Will require a memcopy of u8's across buffers in order to create 1 contiguous buffer
+            // to return.
+            // Removes expended buffers after taking
+
+            let mut out_bytes = Vec::with_capacity(self.size_in_bytes as usize);
+            // this must loop through the bytes buffers because we don't know how many buffers will
+            // be needed to fulfill `size_in_bytes`. e.g. if every buffer had length 1, we'd need to
+            // visit `size_in_bytes` buffers in order to fill `out_bytes`
+            loop {
+                let remaining_bytes = out_bytes.capacity() - out_bytes.len();
+                if remaining_bytes == 0 {
+                    break;
+                }
+                // current `buffer` has more bytes than needed
+                if self.buffered[0].len() > remaining_bytes {
+                    let remaining = self.buffered[0].split_to(remaining_bytes);
+                    out_bytes.extend_from_slice(&remaining);
+                } else {
+                    // current `buffer` has <= bytes needed, remove and append to out_bytes
+                    out_bytes.extend_from_slice(&self.buffered.pop_front().unwrap());
```

Comment:
You should be able to use append for this if this is a vec.  For an array, you will need to assign:

```rust
let buf = self.buffered.pop_front();
out_bytes[next..next+buf.len()]  = buf[..];
```

Reply:
`self.buffered.pop_front()` in this context is a  `hyper::body::Bytes`, which can be borrowed as a slice, but isn't a `Vec`, so I'm not sure how to use `append` for that.

is there a reason to avoid `extend_from_slice`?

Reply:
Only that it assumes you are doing heap allocations.